srtcore/common.cpp

```diff
@@ -380,83 +388,139 @@ void CIPAddress::ntop(const sockaddr_any& addr, uint32_t ip[4])
-          // [00/00/00/00/00/00/00/00/00/00/FF/FF/192/168/128/2]
-          // 32-bit
-          // [00000000 && 00000000 && 0000FFFF && 192.168.128.2]
-
-          // Spreading every 16-bit word separately to avoid endian dilemmas
-          paddr16[2 * 2 + 0] = 0;
-          paddr16[2 * 2 + 1] = 0xFFFF;
-
-          paddr32[3] = ip[0]; // IPv4 address encoded here
-      }
-   }
+        // This whole above procedure was only in order to EXCLUDE the
+        // possibilirt of IPv4-mapped-on-IPv6. This below may only happen
+        // if BOTH peers are IPv6. Otherwise we have a situation of cross-IP
+        // version connection in which case the address in question is always
+        // IPv4 in various mapping formats.
+        if (!is_mapped_ipv4)
+        {
+            // Here both agent and peer use IPv6, in which case
+            // `ip` contains the full IPv6 address, so just copy
+            // it as is.
+
+            // XXX Possibly, a simple
+            // memcpy( (a->sin6_addr.s6_addr), ip, 16);
+            // would do the same thing, and faster. The address in `ip`,
+            // even though coded here as uint32_t, is still big endian.
+            for (int i = 0; i < 4; ++ i)
+            {
+                a->sin6_addr.s6_addr[i * 4 + 0] = ip[i] & 0xFF;
+                a->sin6_addr.s6_addr[i * 4 + 1] = (unsigned char)((ip[i] & 0xFF00) >> 8);
+                a->sin6_addr.s6_addr[i * 4 + 2] = (unsigned char)((ip[i] & 0xFF0000) >> 16);
+                a->sin6_addr.s6_addr[i * 4 + 3] = (unsigned char)((ip[i] & 0xFF000000) >> 24);
+            }
+            return; // The address is written, nothing left to do.
+        }
+
+        // 
+        // IPv4 mapped on IPv6
+
+        // Here agent uses IPv6 with IPPROTO_IPV6/IPV6_V6ONLY == 0
+        // In this case, the address in `ip` is always an IPv4,
+        // although we are not certain as to whether it's using the
+        // IPv6 encoding (0::FFFF:IPv4) or SRT encoding (IPv4::0);
+        // this must be extra determined.
+        //
+        // Unfortunately, sockaddr_in6 doesn't give any straightforward
+        // method for it, although the official size of a single element
+        // of the IPv6 address is 16-bit.
+
+        memset((a->sin6_addr.s6_addr), 0, sizeof a->sin6_addr.s6_addr);
+
+        // There are also available sin6_addr.s6_addr32, but
+        // this is kinda nonportable.
+        uint32_t* paddr32 = (uint32_t*)a->sin6_addr.s6_addr;
+        uint16_t* paddr16 = (uint16_t*)a->sin6_addr.s6_addr;
+
+        // layout: of IPv4 address 192.168.128.2
+        // 16-bit:
+        // [0000: 0000: 0000: 0000: 0000: FFFF: 192.168:128.2]
+        // 8-bit
+        // [00/00/00/00/00/00/00/00/00/00/FF/FF/192/168/128/2]
+        // 32-bit
+        // [00000000 && 00000000 && 0000FFFF && 192.168.128.2]
+
+        // Spreading every 16-bit word separately to avoid endian dilemmas
+        paddr16[2 * 2 + 0] = 0;
+        paddr16[2 * 2 + 1] = 0xFFFF;
```

Comment:
`paddr16[2 * 2 + 0] = 0;` ? There was memset to zero right before.

Reply:
Might be that it wasn't memset for the full address before, now that it is, indeed this isn't necessary.